### PR TITLE
Collector slabinfo

### PR DESCRIPTION
**Pull request by Saruspete**
#### Summary

Provide new collector parsing `/proc/slabinfo` to provide details on kernel slab structures.

Asked by issue #13 (very happy for the oldest issue in backlog) 

##### Component Name

collectors/slabinfo.plugin

##### Additional Information

This slabinfo details allows to have clues on actions done on your system.
In the following screenshot, you can clearly see a `find` done on a ext4 filesystem (the number of `ext4_inode_cache` & `dentry` are rising fast), and a few seconds later, an admin issued a `echo 3 > /proc/sys/vm/drop_cached` as their count dropped.

netdata_slabinfo

### CODE REVIEW

**Comment by vlvkobal** [comment]
I think you should set permissions for the plugin in the https://github.com/netdata/netdata/blob/master/netdata-installer.sh#L718-#L768 section and mention in the documentation the fact that the plugin has the special privileges and its cause.

**Comment by Saruspete** [comment]
Added link in doc generator & more detail on why we need it setuid (or setcap).

File: collectors/slabinfo.plugin/README.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+Each internal structure (process, file descriptor, inode...) is stored within a SLAB.
+
+
+## configuring Netdata for slabinfo
+
+There is currently no configuration needed, aside to have the `slabinfo.plugin` plugin being able to read 
+`/proc/slabinfo`. As this procfile is only readable by root, (mode `0400` in the kernel source), we need to grant the 
+slabinfo.plugin this access, either by:
+-  with Setuid root: `chown root: slabinfo.plugin && chmod u+s slabinfo.plugin`
+-  with capability dac_override: `setcap cap_dac_override+ep slabinfo.plugin`
```
Comment (vlvkobal) [inline collectors/slabinfo.plugin/README.md]: You've already granted access to the `slabinfo` file at installation time. I don't think users need to know about other possible ways of configuring it. You can use a simple sentence, something like
> Keep in mind that libxenstat requires root access, so the plugin is setuid to root.

Comment (Saruspete) [reply collectors/slabinfo.plugin/README.md]: Is that better ? 

> There is currently no configuration needed.
>
> As `/proc/slabinfo` is only readable by root, this plugin needs to be setuid root.

Comment (vlvkobal) [reply collectors/slabinfo.plugin/README.md]: LGTM

**Comment by vlvkobal** [comment]
Please add the plugin to
- `collectors/README.md`
- `collectors/apps.plugin/apps_groups.conf`
- `collectors/plugins.d/README.md`
- `docs/Add-more-charts-to-netdata.md`

You should set `SLABINFO_PLUGIN_FILES` in `CMakeLists.txt`. Please make sure that `make` after `cmake` work correctly.

File: collectors/slabinfo.plugin/slabinfo.c [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+#define SLAB_PAGE_SIZE PAGE_SIZE
+
+// ----------------------------------------------------------------------------
+
+// callback required by fatal()
+void netdata_cleanup_and_exit(int ret) {
+	exit(ret);
+}
+
+void send_statistics( const char *action, const char *action_result, const char *action_data) {
```
Comment (vlvkobal) [inline collectors/slabinfo.plugin/slabinfo.c]: An extra space.

File: collectors/slabinfo.plugin/slabinfo.c [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+
+
+
+unsigned int do_slab_stats(int update_every) {
+
+	static unsigned int loops = 0;
+	struct slabinfo *sactive = NULL, *s = NULL;
+
+	// Main processing loop
+	while ( running ) {
```
Comment (vlvkobal) [inline collectors/slabinfo.plugin/slabinfo.c]: Extra spaces.

**Comment by Saruspete** [comment]
tested cmake, seems ok now.

File: collectors/slabinfo.plugin/slabinfo.c (hunk 3 above)
Comment (Saruspete) [reply collectors/slabinfo.plugin/slabinfo.c]: Fixed. Is there any linter that could check this automatically ?

**Comment by Saruspete** [comment]
For the alarms, I don't think so. Or at least not for the behavior I saw.
Slabinfo is an insight on the kernel memory management, and while it provide a useful way of checking what behavior is consuming kernel resources (file-descriptor (filp), processes (task_struct) fs (dentry, {fuse,rpc,fat,mqueue,ext4,sock,shmem,proc}_inode_cache), network (conntrack, iptables...)) I don't see any generic way of knowing if allocations are an issue or totally relevant for the kind of job done by the server. 

I did use the debug a lot, but removed it as I didn't know how to impelement it or what value to choose in log.h.
do you have any hint on this, so I could put them back ?

**Comment by Saruspete** [comment]
Or for the debug, do you mean I didn't use it correctly ?

File: collectors/slabinfo.plugin/slabinfo.c (hunk 3 above)
Comment (vlvkobal) [reply collectors/slabinfo.plugin/slabinfo.c]: #6677 is on its way to being merged.

**Comment by vlvkobal** [comment]
I meant that usually, we use the `debug` command-line option for an external plugin and simply print debug messages to `stderr`:
```
        else if(strcmp("debug", argv[i]) == 0) {
            debug = 1;
            continue;
        }
```

```
                    " Available command line options:\n"
                    "\n"
                    "  SECONDS                 data collection frequency\n"
                    "                          minimum: %d\n"
                    "\n"
                    "  debug                   enable verbose output\n"
                    "                          default: disabled\n"
```

```
if(debug) fprintf(stderr, "freeipmi.plugin: hostname set to '%s'\n", hostname);
```
It allows debugging the plugin as a separate unit.

**Comment by vlvkobal** [comment]
One last thing - we don't use tabs in C code. Please convert them to spaces (4 spaces per tab).

**Comment by Saruspete** [comment]
Fixed. 
fwiw I'm really not fond of using space (a visual indent) to reflect a logical indent, but unity is more important :)

**Comment by vlvkobal** [comment]
You've introduced compile warnings with the commit for the debug functions:
```
collectors/slabinfo.plugin/slabinfo.c: In function 'get_slabstruct':
collectors/slabinfo.plugin/slabinfo.c:22:21: warning: format '%lu' expects argument of type 'long unsigned int', but argument 3 has type 'int' [-Wformat=]
   22 |     fprintf(stderr, "slabinfo.plugin DEBUG (%04lu@%-10.10s:%-15.15s)::", __LINE__, __FILE__, __FUNCTION__); \
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
collectors/slabinfo.plugin/slabinfo.c:22:21: note: in definition of macro 'slabdebug'
   22 |     fprintf(stderr, "slabinfo.plugin DEBUG (%04lu@%-10.10s:%-15.15s)::", __LINE__, __FILE__, __FUNCTION__); \
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
collectors/slabinfo.plugin/slabinfo.c:22:49: note: format string is defined here
   22 |     fprintf(stderr, "slabinfo.plugin DEBUG (%04lu@%-10.10s:%-15.15s)::", __LINE__, __FILE__, __FUNCTION__); \
      |                                             ~~~~^
      |                                                 |
      |                                                 long unsigned int
      |                                             %04u
```

**Comment by Saruspete** [comment]
Well, from the error message, I would say cadacy is having a bad time :
```
%lu in format string (no. 1) requires 'unsigned long' but the argument type is 'size_t {aka unsigned long}'.
```

Is that a case of "error success" ?

**Comment by mfundul** [comment]
> 
> 
> Well, from the error message, I would say cadacy is having a bad time :
> 
> ```
> %lu in format string (no. 1) requires 'unsigned long' but the argument type is 'size_t {aka unsigned long}'.
> ```
> 
> He wants `unsigned long`, but the type is `unsigned long` Is that a case of "error success" ?

You're not going to like what I am about to say, but you have to cast it to `(unsigned long)` to make the compiler happy, and we need the project to not have any warnings. The explanation is something like that for the compiler it has been defined as a different type even though it's a `typedef` of identical properties.

**Comment by Saruspete** [comment]
And on travis, the error is not related to the build : 
```
tests/updater_checks.sh: line 80: bats: command not found

The command "docker run -it -v "${PWD}:/netdata:rw" -w /netdata "centos:7" tests/updater_checks.sh" exited with 127.
```

**Comment by vlvkobal** [comment]
> And on travis, the error is not related to the build

It doesn't mean you should ignore the Coverity issues ;)

**Comment by Saruspete** [comment]
Cast done. will retrigger travis to.

**Comment by Saruspete** [comment]
Force pushed to resolve conflict in `netdata.spec.in` due to cups plugin added where slabinfo.plugin was.

What's the usual wait-time for merge ? Does it need to be part of a sprint or milestone ? (asking for PR #6843 too ^^)

**Comment by vlvkobal** [comment]
It looks like you force pushed the old version. Several last commits have disappeared.

**Comment by Saruspete** [comment]
Thankfully you have an eagle eye !
The more mistakes I make, the more I believe all of you are trustworthy to keep netdata clean and stable !

**Comment by cakrit** [comment]
> What's the usual wait-time for merge ? Does it need to be part of a sprint or milestone ? (asking for PR #6843 too ^^)

It only depends on how long it takes us to do a proper review. PRs for C code addition/changes are always trickier, but we generally try to prioritize bug fixes and PR reviews over other work.

**Comment by Saruspete** [comment]
Ok makes sense :-) thanks for the info !

**Pull request merged**